opentracing/propagation.go

```diff
@@ -0,0 +1,299 @@
+package opentracing
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/binary"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+
+	opentracing "github.com/opentracing/opentracing-go"
+)
+
+const (
+	fieldNameTraceID = "traceid"
+	fieldNameSpanID  = "spanid"
+	fieldNameSampled = "sampled"
+)
+
+type splitTextPropagator struct {
+	tracer *Tracer
+}
+
+type splitBinaryPropagator struct {
+	tracer *Tracer
+}
+
+type goHTTPPropagator struct {
+	*splitBinaryPropagator
+}
+
+func (p *splitTextPropagator) InjectSpan(
+	sp opentracing.Span,
+	carrier interface{},
+) error {
+	sc := sp.(*Span)
+	splitTextCarrier, ok := carrier.(*opentracing.SplitTextCarrier)
+	if !ok {
+		return opentracing.ErrInvalidCarrier
+	}
+	splitTextCarrier.TracerState = map[string]string{
+		fieldNameTraceID: strconv.FormatUint(uint64(sc.Recorder.SpanID.Trace), 10),
+		fieldNameSpanID:  strconv.FormatUint(uint64(sc.Recorder.SpanID.Span), 10),
+		fieldNameSampled: strconv.FormatBool(sc.sampled),
+	}
+	sc.Lock()
+	splitTextCarrier.TraceAttributes = make(map[string]string, len(sc.attributes))
+	for k, v := range sc.attributes {
+		splitTextCarrier.TraceAttributes[k] = v
+	}
+	sc.Unlock()
+	return nil
+}
+
+func (p *splitTextPropagator) JoinTrace(
+	operationName string,
+	carrier interface{},
+) (opentracing.Span, error) {
+	splitTextCarrier, ok := carrier.(*opentracing.SplitTextCarrier)
+	if !ok {
+		return nil, opentracing.ErrInvalidCarrier
+	}
+	requiredFieldCount := 0
+	var traceID, propagatedSpanID uint64
+	var sampled bool
+	var err error
+	for k, v := range splitTextCarrier.TracerState {
+		switch strings.ToLower(k) {
+		case fieldNameTraceID:
+			traceID, err = strconv.ParseUint(v, 10, 64)
+			if err != nil {
+				return nil, opentracing.ErrTraceCorrupted
+			}
+			requiredFieldCount++
+		case fieldNameSpanID:
+			propagatedSpanID, err = strconv.ParseUint(v, 10, 64)
+			if err != nil {
+				return nil, opentracing.ErrTraceCorrupted
+			}
+			requiredFieldCount++
+		case fieldNameSampled:
+			sampled, err = strconv.ParseBool(v)
+			if err != nil {
+				return nil, opentracing.ErrTraceCorrupted
+			}
+			requiredFieldCount++
+		default:
+			return nil, fmt.Errorf("Unknown TracerState field: %v", k)
+		}
+	}
+
+	if requiredFieldCount < 3 {
+		return nil, fmt.Errorf("Only found %v of 3 required fields", requiredFieldCount)
```

Comment:
`Only` should not start with a capital `O` according to https://github.com/golang/go/wiki/CodeReviewComments#error-strings